xain_fl/config/schema.py

```diff
@@ -0,0 +1,389 @@
+       bucket = "aggregated_weights"
+
+       # AWS secret access to use to authenticate to the storage service
+       secret_access_key = "my-secret"
+
+       # AWS access key ID to use to authenticate to the storage service
+       access_key_id = "my-key-id"
+
+    This config file can be loaded and used as follows:
+
+    .. code-block:: python
+
+       from xain_fl.config import Config
+
+       config = Config.load("example_config.toml")
+
+       assert config.server.host == "[::]"
+       assert config.server.port == 50051
+
+       assert config.ai.initial_weights == "./test_array.npy"
+       assert config.ai.rounds == 1
+       assert config.ai.epochs == 1
+       assert config.ai.min_participants == 1
+       assert config.ai.fraction_participants == 1.0
+
+       assert config.storage.endpoint == "http://localhost:9000"
+       assert config.storage.bucket == "aggregated_weights"
+       assert config.storage.secret_access_key == "my-access-key"
+       assert config.storage.access_key_id == "my-key"
+    """
+
+    def __init__(self, ai: NamedTuple, storage: NamedTuple, server: NamedTuple):
+        self.ai = ai
+        self.storage = storage
+        self.server = server
+
+    @classmethod
+    def from_unchecked_dict(cls: Type[T], dictionary: Mapping[str, Any]) -> T:
+        """Check if the given dictionary contains a valid configuration, and
+        if so, create a `Config` instance from it.
+
+        Args:
+
+            dictionary: a dictionary containing the configuration
+        """
+        try:
+            valid_config = CONFIG_SCHEMA.validate(dictionary)
+        except SchemaError as err:
+            raise InvalidConfig(err.code)
+        return cls.from_valid_dict(valid_config)
+
+    @classmethod
+    def from_valid_dict(cls: Type[T], dictionary: Mapping[str, Any]) -> T:
+        """Create a `Config` instance for the given dictionary, assuming it
+        contains a valid configuration
+
+        Args:
+
+            dictionary: a dictionary containing the configuration
+
+        """
+        return cls(
+            AiConfig(**dictionary["ai"]),
+            StorageConfig(**dictionary["storage"]),
+            ServerConfig(**dictionary["server"]),
+        )
+
+    @classmethod
+    def load(cls: Type[T], path: str) -> T:
+        """Read the config file from the given path, check that it contains a
+        valid configuration, and return the corresponding `Config`
+        instance.
+
+        Args:
+
+            path: path to a configuration file
+        """
+        try:
+            with open(path, "r") as f:
+                raw_config = toml.load(f)
+        except IsADirectoryError:
+            raise InvalidConfig(f"{path} is a directory")
+        except FileNotFoundError:
+            raise InvalidConfig(f"{path} not found")
+        except PermissionError:
+            raise InvalidConfig(f"failed to read {path}: insufficient permissions")
```

Comment:
in case if we would like to have some information in readable form, would be nice to log the error here and have more generic exception message.

Reply:
I believe your referring to [my earlier comment](https://github.com/xainag/xain-fl/pull/221#discussion_r368863619)? If so I ended up changing my mind about this actually :p Logging these exceptions when we catch them seems to print clear enough logs. Here is an example: https://github.com/xainag/xain-fl/pull/221#discussion_r369185286

What do you think?